### PR TITLE
CheckLabelShaped to check and correct label shapes

### PR DESCRIPTION
**Pull request by myron**
adds CheckLabelShaped transform under monai.apps.auto3dseg:

- to check if image shape and label shape are the same. Currently, we usually don't check for it, which results in potentially a non-descriptive crash later, when computing loss or metric. 
- to correct label shape if it is very similar to image shape. Unfortunately some datasets, such as Hecktor22, has some labels (masks) with  a slightly different shapes (difference in only 1 voxel). And because of that a user will have to either re-save the whole dataset, or account for it during the training loop (which is inconvenient). this utility auto corrects it.

We also already have this auto-correction in DataAnalyzer, and with this transform, DataAnalyzer logic is simplified.

### CODE REVIEW

File: monai/apps/auto3dseg/transforms.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+from monai.networks.utils import pytorch_after
+from monai.transforms import MapTransform
+
+
+class CheckLabelShaped(MapTransform):
+    """
+    Checks if segmentation label images (in keys) have the same spatial shape as the main image (in source_key),
+    and raise an error if the shapes are significantly different.
+    if the shapes are only slightly different (within an allowed_shape_difference in each dim), then quietly resize the label.
+    This transform is designed to quietly correct datasets with label shape mismatches. Generally image and segmentation label
```
Comment (mingxin-zheng) [inline monai/apps/auto3dseg/transforms.py:27]: We can mention the function and method used for the correction in the doc-string.

File: monai/apps/auto3dseg/transforms.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    must have the same spatial shape, however some public datasets having a slight shape mismatches, which will cause
+    potential crashes when calculating loss or metric functions.
```
Comment (mingxin-zheng) [inline monai/apps/auto3dseg/transforms.py:29]: Please add the doc-string for the input arguments. Thanks!

File: monai/apps/auto3dseg/transforms.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        allow_missing_keys: bool = False,
+        source_key: str = "image",
+        allowed_shape_difference: int = 5,
+    ) -> None:
+
+        super().__init__(keys=keys, allow_missing_keys=allow_missing_keys)
+        self.source_key = source_key
+        self.allowed_shape_difference = allowed_shape_difference
+
+    def __call__(self, data: Mapping[Hashable, torch.Tensor]) -> Dict[Hashable, torch.Tensor]:
```
Comment (mingxin-zheng) [inline monai/apps/auto3dseg/transforms.py:44]: As mentioned above, a docstring of the input argument would be necessary.

**Comment by mingxin-zheng** [comment]
Hi @myron thank you for the PR and having me to review it. 

I have a question for the other metadata in the image, such as the `affine` and `spacing`. Since there is an interpolation, I feel we need to adjust those accordingly. Otherwise the datastats results of these metadata are incorrect in subsequent steps.

File: monai/apps/auto3dseg/transforms.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+from monai.config import KeysCollection
+from monai.networks.utils import pytorch_after
+from monai.transforms import MapTransform
+
+
+class CheckLabelShaped(MapTransform):
+    """
+    Checks if segmentation label images (in keys) have the same spatial shape as the main image (in source_key),
+    and raise an error if the shapes are significantly different.
+    if the shapes are only slightly different (within an allowed_shape_difference in each dim), then quietly resize the label.
```
Comment (mingxin-zheng) [inline monai/apps/auto3dseg/transforms.py:26]: May I ask why it should be done quietly?

Comment (wyli) [reply monai/apps/auto3dseg/transforms.py:26]: maybe the transform's name should be EnsureSameShape or EnsureShape...

File: monai/apps/auto3dseg/transforms.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    must have the same spatial shape, however some public datasets having a slight shape mismatches, which will cause
+    potential crashes when calculating loss or metric functions.
+    """
+
+    def __init__(
+        self,
+        keys: KeysCollection = "label",
+        allow_missing_keys: bool = False,
+        source_key: str = "image",
+        allowed_shape_difference: int = 5,
```
[the remark this answers is not in the document]
Comment (myron) [reply monai/apps/auto3dseg/transforms.py:37]: let's just use 5 everywhere

File: monai/apps/auto3dseg/transforms.py (hunk 2 above)
Comment (myron) [reply monai/apps/auto3dseg/transforms.py:29]: okay

File: monai/apps/auto3dseg/transforms.py (hunk 1 above)
Comment (myron) [reply monai/apps/auto3dseg/transforms.py:27]: ok

File: monai/apps/auto3dseg/transforms.py (hunk 4 above)
Comment (myron) [reply monai/apps/auto3dseg/transforms.py]: > May I ask why it should be done quietly?

it's designed to correct a tiny shape mismatch. We could show a message or warning, but this can  potentially print many statements for the user.. if you think we need it , let me know

**Comment by myron** [comment]
> he other metadata in the image, such as the `affine` and `spacing`. Since there is an interp

it's designed to correct a tiny shape mismatch, due to some datasets are incorrectly saved. The correction is just a view voxels, I don't think we need to update affine or spacing.

**Comment by myron** [comment]
> May I ask why it should be done quietly?
@mingxin-zheng it's designed to correct a tiny shape mismatch. We could show a message or warning, but this can potentially print many statements for the user.. if you think we need it , let me know

**Comment by myron** [comment]
the windows test fails (due to some unrelated reason), @wyli can you take a look plz

`C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\torch\include\pybind11\detail/common.h(152): warning C4005: 'HAVE_SNPRINTF': macro redefinition
C:\hostedtoolcache\windows\Python\3.8.10\x64\include\pyerrors.h(315): note: see previous definition of 'HAVE_SNPRINTF'
creating D:\a\MONAI\MONAI\build\lib.win-amd64-3.8
creating D:\a\MONAI\MONAI\build\lib.win-amd64-3.8\monai
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\torch\lib /LIBPATH:C:\hostedtoolcache\windows\Python\3.8.10\x64\libs /LIBPATH:C:\hostedtoolcache\windows\Python\3.8.10\x64\PCbuild\amd64 "/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22621.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\\lib\10.0.22621.0\\um\x64" c10.lib torch.lib torch_cpu.lib torch_python.lib /EXPORT:PyInit__C D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\ext.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\filtering\bilateral\bilateral.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\filtering\bilateral\bilateralfilter_cpu.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\filtering\bilateral\bilateralfilter_cpu_phl.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\filtering\permutohedral\permutohedral.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\filtering\permutohedral\permutohedral_cpu.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\lltm\lltm_cpu.obj D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\resample\pushpull_cpu.obj /OUT:build\lib.win-amd64-3.8\monai\_C.pyd /IMPLIB:D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\_C.lib /openmp
LINK : warning LNK4044: unrecognized option '/openmp'; ignored
   Creating library D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\_C.lib and object D:\a\MONAI\MONAI\build\temp.win-amd64-3.8\Release\a\MONAI\MONAI\monai\csrc\_C.exp
ext.obj : error LNK2001: unresolved external symbol __imp___tls_offset_?init@?1??lazy_init_num_threads@internal@at@@YAXXZ@4_NA
ext.obj : error LNK2001: unresolved external symbol __imp___tls_index_?init@?1??lazy_init_num_threads@internal@at@@YAXXZ@4_NA
build\lib.win-amd64-3.8\monai\_C.pyd : fatal error LNK1120: 2 unresolved externals
error: command 'C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.34.31933\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
Error: Process completed with exit code 1.
`

File: monai/auto3dseg/analyzer.py [hunk 6]
```diff
@@ -465,16 +455,7 @@ def __call__(self, data):
...
-            if np.allclose(list(ndas_label.shape), list(ndas[0].shape), atol=10):
-                logger.info(f" Label shape {ndas_label.shape} is slightly different from image shape {ndas[0].shape}")
-                ndas_label = F.interpolate(
-                    input=ndas_label.unsqueeze(0).unsqueeze(0),
-                    size=list(ndas[0].shape),
-                    mode="nearest-exact" if pytorch_after(1, 11) else "nearest",
-                )[0, 0]
-            else:
-                raise ValueError(f"Label shape {ndas_label.shape} is  different from image shape {ndas[0].shape}")
+            raise ValueError(f"Label shape {ndas_label.shape} is  different from image shape {ndas[0].shape}")
```
Comment (mingxin-zheng) [inline monai/auto3dseg/analyzer.py:458]: There are two spaces between "is" and "different" here

**Comment by mingxin-zheng** [comment]
Hi @myron , can you also update the docstring in DataAnalyzer about the shape check? Since the difference is hardcoded to be 5 now, I think the users need to be informed about this too.

File: monai/auto3dseg/analyzer.py (hunk 6 above)
Comment (myron) [reply monai/auto3dseg/analyzer.py:458]: ok

**Comment by myron** [comment]
> Hi @myron , can you also update the docstring in DataAnalyzer about the shape check? Since the difference is hardcoded to be 5 now, I think the users need to be informed about this too.

ok, I added more doc strings, and made it an options for a user to change @mingxin-zheng  plz check

**Review by mingxin-zheng — APPROVED** [review]

File: monai/apps/auto3dseg/transforms.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        self.allowed_shape_difference = allowed_shape_difference
+
+    def __call__(self, data: Mapping[Hashable, torch.Tensor]) -> Dict[Hashable, torch.Tensor]:
+        d = dict(data)
+        image_shape = d[self.source_key].shape[1:]
+        for key in self.key_iterator(d):
+            label_shape = d[key].shape[1:]
+            if label_shape != image_shape:
+                if np.allclose(list(label_shape), list(image_shape), atol=self.allowed_shape_difference):
+                    warnings.warn(f"The {key} shape {label_shape} is different from the source shape {image_shape}.")
```
Comment (wyli) [inline monai/apps/auto3dseg/transforms.py:60]: should mention the shape is updated from x to y and the metadata is not updated

Comment (myron) [reply monai/apps/auto3dseg/transforms.py]: okay, I've updated the warning message.
